### PR TITLE
Prevent expanding instructor feedback result pages during testing #6981

### LINKED ISSUE #6981: Prevent expanding instructor feedback result pages during testing
**Description of feature/enhancement**
A follow up issue for #6608 .  Currently all instructor feedback result page panels are expanded during testing and it slows down the UI testing.

**Justification**
The expansion of panels is unnecessary in many test cases in InstructorFeedbackResultPageUiTest.java and it should be removed to save up test running time.

### PR DESCRIPTION
**Pull request by boxin-yang**
Fixes #6981 

**Outline of Solution**

Prevent expanding all panels as a follow up to #6608 . Right now in instructorFeedbackResultPageUiTest, all the pages expand panels after refreshing. However, to test most functionalities, only one panel expansion is needed and expanding all the panels incur unnecessary waiting time. This PR aims to expand only needed panels and change html content using godmod accordingly.

### CODE REVIEW

**Comment by boxin-yang** [comment]
@junhaoyap @wkurniawan07 
 This is a follow up PR for #6608 . In this PR I do not expand all panels during testing and Only expand panels needed (usually one or two) for testing. Also, I removed some old method such as clickAjaxLoadResponsesPanel , whose name is misleading(this method only clicks panel when the panel ajax class is ajax_submit, which only happens when data loaded is too much) and whose implementation is very inefficient(getting all ajax panel and choose with index).

Also, wait for panel ajax loading in waitForPanelToExpand by waiting for the ajax class to be removed. (This is necessary as panel ajax loading takes time and might cause NoSuchElementException during test)

Ready for review

**Comment by boxin-yang** [comment]
@ablyx Please also take a look when u have time. This is a follow up so might be easier to be reviewed by u haha

**Comment by boxin-yang** [comment]
@junhaoyap @wkurniawan07 ready for review

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java [hunk 1]
```diff
@@ -134,17 +137,19 @@ public void testExceptionalCases() throws Exception {
...
     private void testModerateResponsesButton() {
 
         resultsPage = loginToInstructorFeedbackResultsPage("CFResultsUiT.instr", "Open Session");
         resultsPage.displayByQuestion();
-
+        clickAjaxLoadedPanelAndWaitForExpansion("panelHeading-2", "ajax_auto");
+        clickAjaxLoadedPanelAndWaitForExpansion("panelHeading-4", "ajax_auto");
         ______TS("Typical case: test moderate responses button for individual response (including no response)");
 
+        ThreadHelper.waitFor(1000);
```
Comment (ablyx) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: Why are you using `ThreadHelper` here?

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java [hunk 2]
```diff
@@ -481,32 +500,36 @@ private void testPanelsCollapseExpand() {
...
-        resultsPage.waitForPanelsToCollapse();
+        clickCollapseExpandButtonAndWaitForPanelsToRespond(false);
         assertEquals("Expand Questions", resultsPage.collapseExpandButton.getText());
         assertEquals("Expand all panels. You can also click on the panel heading to toggle each one individually.",
                      resultsPage.collapseExpandButton.getAttribute("data-original-title"));
         resultsPage.verifyResultsHidden();
 
-        resultsPage.clickCollapseExpand();
-        resultsPage.waitForPanelsToExpand();
+        clickCollapseExpandButtonAndWaitForPanelsToRespond(true);
```
Comment (ablyx) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: Is this really needed since we have already expanded all panels at the start?

Comment (boxin-yang) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: That is no longer the case after this PR : [Instructor view results: do not auto-load 3 questions #6568](https://github.com/TEAMMATES/teammates/pull/6608). Now the pages are not auto expanded at the start. The reason I expanded all panels here is because this test is to **testPanelsCollapseExpand** and I think clicking on collapse expand button is needed here.

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java [hunk 3]
```diff
@@ -481,32 +500,36 @@ private void testPanelsCollapseExpand() {
...
         resultsPage.displayByQuestion();
 
         ______TS("Typical case: panels expand/collapse");
 
+        assertEquals("Expand Questions", resultsPage.collapseExpandButton.getText());
+        assertEquals("Expand all panels. You can also click on the panel heading to toggle each one individually.",
+                     resultsPage.collapseExpandButton.getAttribute("data-original-title"));
+        resultsPage.verifyResultsHidden();
+
+        clickCollapseExpandButtonAndWaitForPanelsToRespond(true);
```
Comment (ablyx) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: I mean you have already expanded all panels here, is there a need to test it again below?

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java (hunk 1 above)
Comment (boxin-yang) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: I think this can be safely removed for now as the panel wait function is updated. Thx for the suggestion!

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java (hunk 3 above)
Comment (boxin-yang) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: Erm I think the last check might not be necessary. So the flow now would be:
```    
    collapsed panels => check =>clickToExpandAllPanels
    expanded panels => check => clickToExpandAllPanels
    collapsed panels => check
```
I think this should be enough as it checks and covers all possible states and trasition.

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java [hunk 4]
```diff
@@ -134,15 +137,16 @@ public void testExceptionalCases() throws Exception {
...
+        clickAjaxLoadedPanelAndWaitForExpansion("panelHeading-1", "ajax_auto");
         resultsPage.verifyHtmlMainContent("/instructorFeedbackResultsPageWithSanitizedData.html");
     }
 
     private void testModerateResponsesButton() {
 
         resultsPage = loginToInstructorFeedbackResultsPage("CFResultsUiT.instr", "Open Session");
         resultsPage.displayByQuestion();
-
+        clickAjaxLoadedPanelAndWaitForExpansion("panelHeading-2", "ajax_auto");
```
Comment (ablyx) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: should this be here or should this be under the typical case

Comment (ablyx) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: my bad, looks like in this test case the convention is that set up is done before the test case

Comment (boxin-yang) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: I put it here to follow the convention within this method that environment set up is above _____TS and verifyModerateResponsesButton is below _____TS. But I think I should have added a new line here to avoid confusion!

**Comment by boxin-yang** [comment]
@junhaoyap  Hi, Please take a look when you have time :) In this PR, instead of expanding all panels, which is unnecessarily slow, I expand only panels needed(usually only one panel) for testing. So most of the test would look like this : 
screen shot 2017-04-04 at 5 48 48 pm

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java (hunk 1 above)
Comment (junhaoyap) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: Try not to use ```ThreadHelper``` or waits like that, it would be better to say wait for certain elements to appear instead

File: src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java [hunk 5]
```diff
@@ -68,36 +68,35 @@ private void testInstructorResultsPage() throws Exception {
         InstructorFeedbackResultsPage instructorResultsPage =
                 loginToInstructorFeedbackResultsPageWithViewType("teammates.test.instructor", "openSession2",
                                                                  false, "question");
-        instructorResultsPage.waitForPanelsToExpand();
-
+        clickAjaxLoadedPanelAndWaitForExpansion(instructorResultsPage, "panelHeading-1", "ajax_auto");
```
Comment (junhaoyap) [inline src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java]: I don't think there's a need to test this multiple times, if we test it once and it works, great - especially since an ajax call extends our test time

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 6]
```diff
@@ -189,6 +161,7 @@ public String getQuestionAdditionalInfoButtonText(int qnNumber, String additiona
     }
 
     public void addFeedbackResponseComment(String addResponseCommentId, String commentText) {
+        waitForElementPresence(By.id(addResponseCommentId));
```
Comment (junhaoyap) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: Is this part of the issue?

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 7]
```diff
@@ -364,6 +335,7 @@ public void clickViewPhotoLink(String panelBodyIndex, String urlRegex) {
 
     public void hoverClickAndViewStudentPhotoOnHeading(String panelHeadingIndex, String urlRegex) {
         String idOfPanelHeading = "panelHeading-" + panelHeadingIndex;
+        waitForElementPresence(By.id(idOfPanelHeading));
```
Comment (junhaoyap) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: Is this part of the issue?

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 8]
```diff
@@ -286,6 +256,17 @@ public void waitForPanelsToExpand() {
         }
     }
 
+    public void waitForPanelToExpand(final String panelId, final String ajaxClass) {
+        WebDriverWait wait = new WebDriverWait(browser.driver, TestProperties.TEST_TIMEOUT);
+        wait.until(new ExpectedCondition<Boolean>() {
```
Comment (junhaoyap) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: Please use ```attributeContains``` instead, let's not rewrite what is already there

**Comment by boxin-yang** [comment]
@junhaoyap  I have fixed comments as suggested. However, for the comments on using attributeContains I still do not understand. What I did was to wait until the ajax class is removed, which only happens when ajax content is loaded and panel is clicked. If you suggest testing the attributes of the panel, we still have to use wait.util here

**Comment by boxin-yang** [comment]
@junhaoyap  I have comment and tag u on the possibly redundant part in the test. Please take a look. As for the suggestion to use attributeContains, 
```
wait.until(!ExpectedConditions.attributeContains(
               panelElement, "class", ajaxClass));
```
 is Invalid because until's param is Predicate<Boolean> instead of Boolean. Also, using ExpectedConditions also require importing the class (the class is not imported here). So I think the current implementation might be the best we can do, and I will add method Doc to explain once you think the current solution is ok. (Actually, it is the common way to wait from what I read online)

**Comment by junhaoyap** [comment]
Looking at the documentation -> https://seleniumhq.github.io/selenium/docs/api/java/org/openqa/selenium/support/ui/ExpectedConditions.html

We can do

```
wait.until(ExpectedConditions.not(ExpectedConditions.attributeContains(panelElement, "class", ajaxClass)));
```

**Comment by junhaoyap** [comment]
And there isn't a problem to import the class, we imported ```import org.openqa.selenium.support.ui.ExpectedCondition;``` in order to do what we are doing now, we can change that to ```import org.openqa.selenium.support.ui.ExpectedConditions;```

File: src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java [hunk 9]
```diff
@@ -68,38 +68,8 @@ private void testInstructorResultsPage() throws Exception {
...
         instructorResultsPage.verifyHtmlMainContent("/instructorFeedbackResultsPageRubricQuestionView.html");
-
-        // Giver Recipient Question View
-        instructorResultsPage =
```
Comment (junhaoyap) [inline src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java]: Confirming the removal of these is because they are already being tested in another place, i.e. the unexpanded version while we fix the expected conditions

**Comment by boxin-yang** [comment]
@junhaoyap Just updated the waitForPanelToExpand method with method doc. Also made some changes to better testing. Ready for review

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 10]
```diff
@@ -138,27 +125,11 @@ public void clickShowStats() {
...
-        } catch (NoSuchElementException e) {
-            // Pages with no feedback response do not have button element
-            // We need to click on each of the panels to expand them
-            List<WebElement> panels = browser.driver.findElements(By.className("panel-heading"));
-            for (WebElement panel : panels) {
-                click(panel);
-            }
-            waitForPanelsToExpand();
-        }
+    public void clickElementById(String elementId) {
```
Comment (junhaoyap) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: This method looks like it belongs in AppPage more than here since it is a very generic method, can you check if AppPage has it and if it doesn't, add it to AppPage instead of here

File: src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java (hunk 5 above)
Comment (junhaoyap) [inline src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java]: Add a comment before ```clickAjaxLoadedPanelAndWaitForExpansion``` documenting that this is the test that tests the AJAX functionality so that it won't be accidentally removed

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 11]
```diff
@@ -913,6 +913,11 @@ public AppPage verifyContains(String searchString) {
         return this;
     }
 
+    public void verifyContainElement(By by) {
```
Comment (junhaoyap) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Why did we add this? It isn't used 🤔

File: src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java (hunk 9 above)
Comment (junhaoyap) [inline src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java]: Can we address this? I understand that RGQ, GRQ etc are tested on ```FeedbackRankQuestionUiTest.java``` and I want to make sure that they are meant for testing the same thing and that's why we are removing them..

1. If they are testing the same thing, we can remove them entirely, the name does indicate some difference so we should make sure, one is feedback ui, the other is rubric so we need to investigate if they are the same thing

2. If they are not testing the same thing, the solution here should be to still continue testing all these and remove the AJAX portion, i.e. test the HTML but no need to wait for panels to expand

**Comment by boxin-yang** [comment]
@junhaoyap  
For the deleted part in **FeedbackRubricQuestionUiTest**, the test page itself contains very little data. I highly doubt the necessity of this test because it does not contain any Rubric content in its panel data and that is why I think testing with other sorting views(Giver Recipient Question View etc) is not necessary. Testing with more views adds no value in detecting errors. Ready for review. Please take a look

File: .github/template-helpRequest.md [hunk 12]
```diff
@@ -4,7 +4,7 @@ Copy-and-paste the following template to the issue tracker and fill it in with r
 ```markdown
 - **Operating System:**
 - **JDK Version:**
-- **Eclipse Version:**
+- **Eclipse / IntelliJ Version (if applicable):**
```
Comment (junhaoyap) [inline .github/template-helpRequest.md]: I meant all these changes, I am not sure why they suddenly appeared

**Comment by junhaoyap** [comment]
As discussed on Slack, please make all the weird changes disappear 😂 

Also, revert the changes on the rubric HTML testing and let's tackle the fact that the tests do nothing in a new issue and let's merge the part on the panel expansions first

**Comment by boxin-yang** [comment]
@junhaoyap Ihave reset FeedbackRubricQuestionUiTest back to the initial state(without panels expansion). As we have discussed, the test data used in testing instructorFeedbackResultsPage does not look useful and it is a separate issue. We will open another issue and PR to look further into that. Ready for review

File: src/test/java/teammates/test/pageobjects/AppPage.java (hunk 11 above)
Comment (junhaoyap) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: @boxin-yang still need one last change here before we set it to merge, this method doesn't seem to be used

**Comment by junhaoyap** [comment]
@boxin-yang looks good, can you also make sure to open the issue and link here while you remove the unused method? After that we should be good to go :)

**Comment by boxin-yang** [comment]
@junhaoyap 
 I realized that I made a mistake along our discussion about the testing data in Rubric test. I skipped the test before testInstructorResultsPage() and runs the testInstructorResultsPage() directly and got feedback sessions with almost empty data. However, the data for this test is actually not from test data but rather built up with prior tests in
```
testEditPage();
testInstructorSubmitPage();
testStudentSubmitPage();
testStudentResultsPage();
```

Therefore, the question about useless test data is only because I carelessly ignored the tests before testInstructorResultsPage() and assumed there are independent.

Right now, I open one panel in each of the testInstructorResultsPage test cases for faster page expansion. As for testing for different views, it might be under the consideration that different view used different ajax files so that testing under different views ensures changes with rubric question does not cause failure in all views.

As for verifyContainElement, it is used to replace verifyContains in InstructorFeedbackResultsPageUiTest::testFeedbackResponseCommentActions() ,
```
        ______TS("Typical case: add new feedback response comments");

        resultsPage.displayByRecipientGiverQuestion();
        resultsPage.addFeedbackResponseComment("showResponseCommentAddForm-0-0-1-1", "test comment 1");
        resultsPage.addFeedbackResponseComment("showResponseCommentAddForm-0-0-1-1", "test comment 2");
        resultsPage.verifyCommentRowContent("-0-1-0-1-1", "test comment 1", "CFResultsUiT.instr@gmail.tmt");
        resultsPage.verifyContainsElement(By.id("frComment-visibility-options-trigger-0-1-0-1-1"));
        resultsPage.verifyCommentRowContent("-0-1-0-1-2", "test comment 2", "CFResultsUiT.instr@gmail.tmt");
        resultsPage.verifyContainsElement(By.id("visibility-options-0-1-0-1-2"));
```
This is a digression from main theme of the PR but I feel this is a good change. The old verifyContains searches the whole html string and is very inefficient. The new verifyContainElement searches only DOM with Jquery, which should be a better approach.

**Comment by boxin-yang** [comment]
@junhaoyap As the discussion to reduce running tests on instructorFeedbackResultsPage is irrelevant to this PR, I have opened another issue #7153 to discuss this problem.

**Comment by junhaoyap** [comment]
@boxin-yang I see, it would be best if they were not dependent but we can make do since it is already heavily intertwined and it would take huge amounts of effort in order to untangle them, as discussed, finally found where that method is used 🌞

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 13]
```diff
@@ -286,6 +249,20 @@ public void waitForPanelsToExpand() {
         }
     }
 
+    /**
+     * Wait for Ajax loaded panel to be expanded. The panel is expanded when Ajax loading finishes,
```
Comment (damithc) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: Waits

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java [hunk 14]
```diff
@@ -481,35 +503,34 @@ private void testPanelsCollapseExpand() {
         resultsPage.clickGroupByTeam();
 
         resultsPage.displayByGiverRecipientQuestion();
+        clickAjaxLoadedPanelAndWaitForExpansion("panelHeading-section-0-1", "ajax_auto");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: Seems like this method always take "ajax_auto" or "ajax_submit" as the 2nd parameter. Any way to avoid this repetition?
 Can be a separate PR.

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java [hunk 15]
```diff
@@ -778,4 +799,17 @@ private void verifyQuestionAdditionalInfoExpand(int qnNumber, String additionalI
...
+    }
+
+    private void clickCollapseExpandButtonAndWaitForPanelsToRespond(boolean isExpandingPanels) {
+        resultsPage.clickCollapseExpandButton();
+        if (isExpandingPanels) {
+            resultsPage.waitForPanelsToExpand();
+        } else {
+            resultsPage.waitForPanelsToCollapse();
+        }
+    }
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackResultsPageUiTest.java]: Is it simpler if we create two methods: one to expand and one to collapse?

**Comment by boxin-yang** [comment]
@damithc  Hi Prof, I have left my thoughts on the update on reducing the repitition with ajax_auto and ajax_submit. As for the **clickCollapseExpandButtonAndWaitForPanelsToRespond**, the name might not sound informative and I have separate it into two methods clickCollapseExpandButtonAndWaitForPanelsToExpand() and clickCollapseExpandButtonAndWaitForPanelsToCollapse(). Ready for review
